[Nicholas Koh Zi Lun] iP

Nicholaskoh1 opened the pull request:
# DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Nicholaskoh1 committed:
Add Jeff.java

Nicholaskoh1 committed:
Add printUserInput function

Nicholaskoh1 committed:
Add list function

Nicholaskoh1 committed:
Add Task.java

Nicholaskoh1 committed:
Add task status and description

Nicholaskoh1 committed:
Add DeadlineTask.java

Nicholaskoh1 committed:
Add EventTask.java

Nicholaskoh1 committed:
Add ToDoTask.java

Nicholaskoh1 committed:
Add 3 different task types: ToDo, Deadline, Event

Nicholaskoh1 committed:
Add automated testing

Nicholaskoh1 committed:
Add JeffException.java

Nicholaskoh1 committed:
Add handle errors

Nicholaskoh1 committed:
Add delete function

Nicholaskoh1 committed:
Add java docs

Nicholaskoh1 committed:
Add FileCorruptException.java

Nicholaskoh1 committed:
Update the task files to get the task text file format string

Nicholaskoh1 committed:
Update Jeff.java to save the task list in the data folder

Nicholaskoh1 committed:
Merge branch 'branch-Level-7'

Nicholaskoh1 committed:
Add date and time format to the tasks

Nicholaskoh1 committed:
Merge branch 'branch-Level-8'

Nicholaskoh1 committed:
Add a way to show tasks for a specific date

Nicholaskoh1 committed:
Merge branch 'branch-Level-8'

Nicholaskoh1 committed:
Add more OOP

Nicholaskoh1 committed:
Merge branch 'branch-MoreOOP'

Nicholaskoh1 committed:
Update java files to be in jeff package

Nicholaskoh1 committed:
Merge branch 'branch-Packages'

Nicholaskoh1 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Nicholaskoh1 committed:
Add JUnit tests

Nicholaskoh1 committed:
Add more JUnit tests

Nicholaskoh1 committed:
Merge branch 'branch-JUnit'

Nicholaskoh1 committed:
Update build.gradle to build JAR files

Nicholaskoh1 committed:
Merge branch 'branch-Jar'

Nicholaskoh1 committed:
Add JavaDocs

Nicholaskoh1 committed:
Add more JavaDocs

Nicholaskoh1 committed:
Update java files to follow coding standards

Nicholaskoh1 committed:
Add a way for users to find tasks by search text

Nicholaskoh1 committed:
Merge branch 'branch-A-JavaDoc'

Nicholaskoh1 committed:
Merge branch 'branch-A-CodingStandard'

Nicholaskoh1 committed:
Merge branch 'branch-Level-9'

Nicholaskoh1 committed:
Add checkstyle files

Nicholaskoh1 committed:
Update files to match the checkstyle rules

Nicholaskoh1 committed:
Merge branch 'branch-A-CheckStyle'

Nicholaskoh1 committed:
Add GUI

Nicholaskoh1 committed:
Merge branch 'branch-Level-10'

sandyk0105 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+                boolean isDone = taskParts[1].equals("1");
+                String taskDescription = taskParts[2];
+                Task currentTask = null;
+
```
Good use of variable names!

Nicholaskoh1 committed:
Delete UI testing

Nicholaskoh1 committed:
Update build.gradle

The build.gradle file did not account for the change in the main file
being run for generating jar file (From jeff.Jeff to jeff.gui.Launcher)
when JavaFx was added.

The build.gradle file also did not run assertions when running the main
file.

Updating the build.gradle file to account for these two issues allows
us to generate a jar file for the updated chatbot program and also
run assertions by default.

C5hives commented on src/main/java/jeff/parser/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public static Command parse(String input) throws JeffException {
+        // Filters the input based on the command given
+        if (input.equals("list")) {
+            return new ListCommand(input);
+        } else if (input.equals("bye")) {
+            return new ExitCommand(input);
+        } else if (input.startsWith("mark")) {
+            return new MarkCommand(input);
+        } else if (input.startsWith("unmark")) {
+            return new UnmarkCommand(input);
+        } else if (input.startsWith("delete")) {
+            return new DeleteCommand(input);
+        } else if (input.startsWith("todo") || input.startsWith("deadline") || input.startsWith("event")) {
+            return new AddCommand(input);
+        } else if (input.startsWith("task")) {
+            return new DateCommand(input);
+        } else if (input.startsWith("find")) {
+            return new FindCommand(input);
+        } else {
+            throw new JeffException();
+        }
```
Should the logic of parsing the first word in the string be extracted out?

You can consider creating an Enum and letting said Enum handle the logic of parsing the first word into a relevant command. This way, the chain of if-else statements can be replaced with a switch statement for better readability.

C5hives reviewed (changes requested):
LGTM—the use of OOP makes the project easy to read and understand. 😄

C5hives commented on src/main/java/jeff/exception/JeffException.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Represents an exception for any errors in this package.
+ */
+public class JeffException extends Exception {
+    private final String message;
+
+    /**
+     * Constructor that creates a JeffException object with a specific message.
+     *
+     * @param message that describes the error
+     */
+    public JeffException(String message) {
+        super(message);
+        this.message = message;
+    }
+
+    /**
+     * Constructor that creates a JeffException object with a general message.
+     */
+    public JeffException() {
+        super();
+        this.message = "Sorry, I don't know what this means.";
+    }
+
+    /**
+     * Returns the message of the error.
+     *
+     * @return the error message
+     */
+    @Override
+    public String toString() {
+        return this.message;
+    }
+}
```
I like how you created a custom exception for the chatbot—it'll help with managing current (and future) exceptions in a more precise manner!

C5hives commented on src/main/java/jeff/parser/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public static Command parse(String input) throws JeffException {
+        // Filters the input based on the command given
+        if (input.equals("list")) {
+            return new ListCommand(input);
+        } else if (input.equals("bye")) {
+            return new ExitCommand(input);
+        } else if (input.startsWith("mark")) {
+            return new MarkCommand(input);
+        } else if (input.startsWith("unmark")) {
+            return new UnmarkCommand(input);
+        } else if (input.startsWith("delete")) {
+            return new DeleteCommand(input);
+        } else if (input.startsWith("todo") || input.startsWith("deadline") || input.startsWith("event")) {
+            return new AddCommand(input);
+        } else if (input.startsWith("task")) {
+            return new DateCommand(input);
+        } else if (input.startsWith("find")) {
+            return new FindCommand(input);
+        } else {
+            throw new JeffException();
+        }
```
Should the logic of parsing the first word in the string be extracted out?

You can consider creating an Enum and letting said Enum handle the logic of parsing the first word into a relevant command. This way, the chain of if-else statements can be replaced with a switch statement for better readability.

Nicholaskoh1 replied:
Thanks for pointing it out!

I wanted to prompt the user with the format of the command when they forget to put a spacing after the command word. For example, "findborrow book". If i try to use an Enum and extract the first word, the first word in this case would be "findborrow" instead of "find". This would then prompt a different message instead and thus I didn't want to use Enum for this.

If you have a solution to this, do let me know!

C5hives commented on src/main/java/jeff/parser/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     * @param text Given text.
+     * @return Same text but with a space in front of each line.
+     */
+    public static String prettyText(String text) {
+        // Split the text into lines
+        String[] lines = text.split("\n");
+
+        // Initialise a StringBuilder
+        StringBuilder resultText = new StringBuilder();
+
+        // Add a space in front each line
+        for (String line : lines) {
+            resultText.append(" ").append(line).append("\n");
+        }
+
+        // Convert the StringBuilder back to a String
+        return resultText.toString();
+    }
```
The method name `prettyText` might not be the best way to name this static method.

1. Although the word 'pretty' can be used as a verb, it may not be the most intuitive interpretation of the word ('pretty' can be an adjective to describe the word 'text' as well).
2. Moreover, the method name also may not be most descriptive in telling readers what the function does, which is to add a leading space to every line in the string.

Perhaps you can consider renaming the method to
- `convertToPrettyText()` (method name that starts with a verb) or
- `toPrettyText()` (similar to the `toString()` method name)

to address issue 1, or you could consider something even more descriptive (and possibly overkill), such as:
- `addLeadingSpaceToText()` (very explicit in terms of what the method does)

to address issue 2.

This ultimately boils down to personal preference, but these would be my suggestions. :)

ITZXITZX replied:
I agree with @C5hives.

Here is sounds like 'pretty' is used as a adjective to describe 'Text'.

Nicholaskoh1 replied:
Thanks for pointing it out! I like the suggestion for `addLeadingSpaceToText()` so I would be replacing the function name with this.

C5hives commented on src/main/java/jeff/task/TaskList.java:
```diff
@@ -0,0 +1,195 @@
+ * Represents a list of tasks.
+ */
+public class TaskList {
+    private final ArrayList<Task> tasks = new ArrayList<>();
```
According to the [coding standard](https://se-education.org/guides/conventions/java/index.html#naming), constant names must be all uppercase using underscore to separate words (aka SCREAMING_SNAKE_CASE).

You may want to consider renaming the variable to screaming snake case, or to remove the `final` keyword (I know the Intellj IDE will suggest you to make this variable `final`).

C5hives commented on src/main/java/jeff/task/TaskList.java:
```diff
@@ -0,0 +1,195 @@
+                // Get the task information
+                String[] taskParts = scanner.nextLine().split(" \\| ");
+
+                // Check if the data file is corrupted
+                if (taskParts.length < 3) {
+                    throw new FileCorruptException();
+                }
```
I like the input validation sprinkled throughout the project to defend against the myriad of possible bad inputs. Nice job!

C5hives commented on src/test/java/jeff/command/AddCommandTest.java:
Your tests are very comprehensive. Nice job!

C5hives reviewed (commented):
LGTM—the use of OOP makes the project easy to read and understand. 😄
Ignore my previous PR comment—I had indicated the wrong review category.

ITZXITZX commented on src/main/java/jeff/command/AddCommand.java:
```diff
@@ -0,0 +1,122 @@
+                    targetTask = new DeadlineTask(deadlineContent, Parser.getLocalDateTime(deadlinePeriod));
+                } catch (DateTimeParseException e) {
+                    throw new JeffException(
+                            "The format is wrong! It should be \"deadline xx /by yyyy-mm-dd HH:mm or hh:mm AM/PM\"!");
```
Is this line [too long](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Line%20length%20should%20be%20no%20longer%20than%20120%20chars.)?

Nicholaskoh1 replied:
Thanks for pointing it out! I will be changing it in my next commit.

ITZXITZX commented on src/main/java/jeff/command/AddCommand.java:
```diff
@@ -0,0 +1,122 @@
+                // Check if the format is correct
+                try {
+                    targetTask = new EventTask(
+                            eventContent, Parser.getLocalDateTime(startPeriod), Parser.getLocalDateTime(endPeriod));
```
Will breaking the line here [improve readability](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Place%20line%20break%20to%20improve%20readability)?

Nicholaskoh1 replied:
Yup I agree that I should break the line here to make it easier to read. Thanks for pointing it out!

ITZXITZX commented on src/main/java/jeff/task/Task.java:
```diff
@@ -0,0 +1,122 @@
+     * @param name Given name to check against.
+     * @return true if the task description contains the given name and false otherwise.
+     */
+    public boolean contains(String name) {
```
Is there a clearer name for `contains`? Since its specifically searching within the task description only, and not the other fields.

Nicholaskoh1 replied:
Thanks for pointing it out! I will change it to make it clearer.

ITZXITZX commented on src/main/java/jeff/task/TaskList.java:
```diff
@@ -0,0 +1,195 @@
+     *
+     * @return A list of the file string representation of the tasks.
+     */
+    public List<String> toFileStrings() {
```
Is there a clearer name for `toFileStrings`, since it is returning not a String but a List.

Nicholaskoh1 replied:
Thanks for pointing it out! I will try to change it to make it clearer.

ITZXITZX reviewed (commented):
Overall your code seems good, with some minor cosmetic issues.

Nicholaskoh1 committed:
Update files to improve code quality

The execute methods for most of the Command classes were too long.

Extracting the AddCommand Class to AddToDoCommand Class,
AddDeadlineCommand Class and AddEventCommand Class removes the need
to check for whether the AddCommand is for a To-Do task, Deadline
task or a Event task. Thus shortening the execute method for these
classes.

Most of the Command classes had their execute method shortened by
splitting the code body into several methods. Some of these
methods are created in the class itself while some were created in
the Parser Class instead since it was a common method for the Command
classes.

Nicholaskoh1 committed:
Update files to add assertions

The java files do not have any assertions.

Assertions are added to various parts of the java files.

This allows the program to terminate if the runtime detects any
possible bugs.

Nicholaskoh1 committed:
Merge pull request #1 from Nicholaskoh1/branch-A-Assertions

Update files to add assertions

Nicholaskoh1 committed:
Merge branch 'master' into branch-A-CodeQuality

Branch "branch-A-CodeQuality" uses the master branch before branch
"branch-A-Assertions" was merged into the master branch.

Thus there is a need to merge the master branch into the branch
"branch-A-CodeQuality".

Nicholaskoh1 committed:
Update some java files to fix bugs

The storing of the date and time of the tasks in tasks.txt are in the
wrong format (dd MMM yyyy HH:mm).

Changing the way the dates are stored in the database to this format
(yyyy-MM-dd HH:mm) instead helps solve the issue.

Nicholaskoh1 committed:
Merge pull request #2 from Nicholaskoh1/branch-A-CodeQuality

Merge "branch-A-CodeQuality" branch to master branch

Nicholaskoh1 committed:
Add shorter aliases for command keywords

Users have to type out the full command keyword (Eg. unmark 2), which
can be quite troublesome if the user uses the chatbot a lot.

By allowing for shorter aliases for command keywords (Eg u instead
of unmark), experienced users will be able to type lesser to do
what they want to do with the chatbot.

Nicholaskoh1 committed:
Merge branch 'branch-C-FriendlierSyntax'

Nicholaskoh1 committed:
Add more JUnit tests

Nicholaskoh1 committed:
Match coding style with Checkstyle

Nicholaskoh1 committed:
Merge branch 'branch-A-MoreTesting'

Nicholaskoh1 committed:
Update files to handle unexpected dates

Nicholaskoh1 committed:
Add screenshots of UI

Nicholaskoh1 committed:
Update README

Nicholaskoh1 committed:
Update README.md

Nicholaskoh1 committed:
Update README.md to show images on github.io

Nicholaskoh1 committed:
Add title to GUI

Nicholaskoh1 committed:
Add title to GUI

Nicholaskoh1 committed:
Update Ui.png to exclude bad line wrappings

Nicholaskoh1 committed:
Update README.md
